plonk-hashing/src/poseidon/poseidon.rs

```diff
@@ -0,0 +1,584 @@
+                })
+            }
+            (Some(_), Some(_)) => {
+                unreachable!("currently no one is using this")
+            }
+        }
+    }
+
+    #[cfg(not(feature = "no-optimize"))]
+    fn linear_combination(
+        c: &mut StandardComposer<F, P>,
+        state: &[Self::Field; WIDTH],
+        coeff: impl IntoIterator<Item = Self::ParameterField>,
+    ) -> Self::Field {
+        let coeffs = coeff.into_iter().collect::<Vec<_>>();
+        let mut remaining = WIDTH;
+        let mut index = 0;
+        let mut result: Self::Field;
+        // the first time you have no accumulated result yet, so you can take 3
+        // inputs
+        if remaining < 3 {
+            // this is unlikely, WIDTH is usually at least 3
+            result = c.arithmetic_gate(|g| {
+                g.witness(state[0], state[1], None)
+                    .add(coeffs[0], coeffs[1])
+            });
+            remaining -= 2;
+        } else {
+            result = c.arithmetic_gate(|g| {
+                g.witness(state[index], state[index + 1], None)
+                    .add(coeffs[index], coeffs[index + 1])
+                    .fan_in_3(coeffs[index + 2], state[index + 2])
+            });
+            index += 3;
+            remaining -= 3;
+        }
+
+        // Now you have an accumulated result to carry, so can only take 2
+        // inputs at a time
+        while remaining > 0 {
+            if remaining < 2 {
+                // Accumulate remaining one
+                result = c.arithmetic_gate(|g| {
+                    g.witness(state[index], result, None)
+                        .add(coeffs[index], Self::ParameterField::one())
+                });
+                remaining -= 1;
+            } else {
+                // Accumulate next two
+                result = c.arithmetic_gate(|g| {
+                    g.witness(state[index], state[index + 1], None)
+                        .add(coeffs[index], coeffs[index + 1])
+                        .fan_in_3(Self::ParameterField::one(), result)
+                });
+                index += 2;
+                remaining -= 2;
+            }
+        }
+        result
+    }
+
+    #[cfg(not(feature = "no-optimize"))]
+    fn product_mds_with_sparse_matrix(
+        c: &mut StandardComposer<F, P>,
+        state: &mut [Self::Field; WIDTH],
+        matrix: &SparseMatrix<Self::ParameterField>,
+    ) {
+        let mut result = Self::zeros::<WIDTH>(c);
+
+        result[0] =
+            Self::linear_combination(c, state, matrix.w_hat.iter().cloned());
+        for (j, val) in result.iter_mut().enumerate().skip(1) {
+            // for each j, result[j] = state[j] + state[0] * v_rest[j-1]
+            *val = c.arithmetic_gate(|g| {
+                g.witness(state[0], state[j], None)
+                    .add(matrix.v_rest[j - 1], F::one())
+            });
+        }
+        *state = result;
+    }
+}
+
+trait HashFunction<const WIDTH: usize, COM = ()> {
```

Comment:
This should be arity, not width since the notion of "width" is a poseidon-only thing.

Reply:
Agreed, but see comment below regarding const generic expressions